### PR TITLE
Add handle to multi platform plugin install

### PR DESCRIPTION
**Pull request by otherpirate**
# Add support to multi platform pluging install

### Check list
- [x] Parse JSON fields
- [x] Install multi platforms
- [x] Update tests

Closes [#172](https://github.com/tsuru/tsuru-client/issues/172)

### CODE REVIEW

**Comment by otherpirate** [comment]
@tcarreira could you please review it?

**Comment by tcarreira** [comment]
Hi @otherpirate 
I think there is a misunderstanding about this issue (the issue is not clear enough)

Giving some context: a plugin is an extension of the tsuru-client. 
That means I will only install a plugin for the same os/arch as the tsuru-client itself.

---

From that example in the #172, one would run the following:
```sh
tsuru-plugin-add rpaasv2 "https://github.com/tsuru/rpaas-operator/releases/latest/download/manifest.json" 
```

The tsuru-client would detect its platform, and choose the right URL to install.
Internally, the result would be the same as running (assuming I'm running on `linux/x86_64`):
```sh
tsuru-plugin-add rpaasv2 "https://github.com/tsuru/rpaas-operator/releases/download/v0.33.1/rpaasv2_0.33.1_Linux_x86_64.tar.gz" 
```

---

You may detect the os/arch at runtime using the [runtime](https://pkg.go.dev/runtime#pkg-constants) package:
```go
runtime.GOOS   // the running program's operating system target: one of darwin, freebsd, linux, and so on. 
runtime.GOARCH // the running program's architecture target: one of 386, amd64, arm, s390x, and so on.

// To view possible combinations of GOOS and GOARCH, run "go tool dist list".
```

(after running `go tool dist list`, I'm using those for the manifest on #172. I've edited all examples in the issue, to avoid confusion)

---

The `PluginBundle` is not directly related to this feature, so that code should not be modified.
This feature should be implemented inside `func (c *PluginInstall) Run(... `

A last note, the old behavior must not break ;) 

---

Thank you for your effort.
I think you are in the right direction 👍

**Comment by otherpirate** [comment]
Great description @tcarreira , thank you

I hope now we are ready to go

If there is other concerns, please let me know

File: tsuru/client/plugin.go
Comment (tcarreira) [inline tsuru/client/plugin.go:184]: `BundleMetadata` -> `PluginMetadata`

File: tsuru/client/plugin.go
Comment (tcarreira) [inline tsuru/client/plugin.go:192]: The `BundleManifest` is intended for a completely different feature.
This should be two different types
```go
type BundleManifest struct {
	Plugins []Plugin
}

type PluginManifest struct {
	SchemaVersion  string            `json:"schemaVersion,omitempty"`
	Metadata       BundleMetadata    `json:"metadata,omitempty"`
	UrlPerPlatform map[string]string `json:"urlPerPlatform,omitempty"`
}
```

**Review by tcarreira — CHANGES_REQUESTED** [review]
I mentioned in the last review, but it was a long message ;) 

The `PluginBundle` is not directly related to this feature, so that code should not be modified.
This feature should be implemented inside `func (c *PluginInstall) Run(...` (not `*PluginBundle`)

I did not review the tests, as they will probably change.

File: tsuru/client/plugin.go
Comment (tcarreira) [reply tsuru/client/plugin.go:192]: another context note, the `PluginBundle` is intended for installing multiple plugins with a single url.
This PR is for installing a single plugin, with a single url (when the plugin could be one of many different binaries, declared in a `manifest.json`)

**Comment by otherpirate** [comment]
Giving up, thx @tcarreira

**Comment by tcarreira** [comment]
nooo, don't give up @otherpirate 😉 
If you activate the "Allow edits by maintainers" I could help you finish it.



Of course, if you still want to give up, it's up to you.

**Comment by tcarreira** [comment]
Hey @otherpirate 
After the merge of #191, this PR can be implemented.
Do you want to give it a last try?

what should be done:
- changes go inside `func installPlugin(pluginName, pluginURL string) error`
- nothing should be changed until `http.Get(pluginURL)` (actually, until `defer resp.Body.Close()` referring to that line)
- bind the content of the pluginURL to the new type `PluginManifest`. If it fails, keep the old behavior
- get GOOS and GOARCH from runtime
- get the new URL from that `PluginManifest.UrlPerPlatform[GOOS/GOARCH]`
- recursively call `installPlugin()` with this new pluginURL

To prevent infinite recursion, the `func installPlugin(...)` could take an extra argument: a counter to be checked for recursion

Let me know either decision ;)

**Pull request closed without merging**
